pumps/resurface.go

```diff
@@ -0,0 +1,217 @@
+package pumps
+
+import (
+	"context"
+	"encoding/base64"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/TykTechnologies/tyk-pump/analytics"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"github.com/resurfaceio/logger-go/v3"
+)
+
+type ResurfacePump struct {
+	logger *logger.HttpLogger
+	config *ResurfacePumpConfig
+	CommonPumpConfig
+}
+
+type ResurfacePumpConfig struct {
+	EnvPrefix string `mapstructure:"meta_env_prefix"`
+	Url       string `mapstructure:"capture_url"`
+	Rules     string
+	Queue     []string
+}
+
+const (
+	resurfacePrefix     = "resurface-pump"
+	resurfacePumpName   = "Resurface Pump"
+	resurfaceDefaultEnv = PUMPS_ENV_PREFIX + "_RESURFACEIO" + PUMPS_ENV_META_PREFIX
+)
+
+func (rp *ResurfacePump) New() Pump {
+	newPump := ResurfacePump{}
+	return &newPump
+}
+
+func (rp *ResurfacePump) GetName() string {
+	return resurfacePumpName
+}
+
+func (rp *ResurfacePump) GetEnvPrefix() string {
+	return rp.config.EnvPrefix
+}
+
+func (rp *ResurfacePump) Init(config interface{}) error {
+	rp.config = &ResurfacePumpConfig{}
+	rp.log = log.WithField("prefix", resurfacePrefix)
+
+	err := mapstructure.Decode(config, &rp.config)
+	if err != nil {
+		rp.log.Fatal("Failed to decode configuration: ", err)
+	}
+
+	processPumpEnvVars(rp, rp.log, rp.config, resurfaceDefaultEnv)
+
+	opt := logger.Options{
+		Rules: rp.config.Rules,
+		Url:   rp.config.Url,
+		Queue: rp.config.Queue,
+	}
+	logger, err := logger.NewHttpLogger(opt)
+	if err != nil {
+		log.Fatal(err)
+	}
+	rp.logger = logger
+	if logger.Enabled() {
+		rp.log.Info(rp.GetName() + " Initialized")
+		return nil
+	} else {
+		rp.log.Info(rp.GetName() + " Initialized (Logger disabled)")
+		return errors.New("Logger is not enabled")
+	}
+}
+
+func parseHeaders(headersString string, existingHeaders http.Header) http.Header {
+	var headers http.Header
+	if existingHeaders != nil {
+		headers = http.Header.Clone(existingHeaders)
+	} else {
+		headers = http.Header{}
+	}
+	for _, line := range strings.Split(headersString, "\r\n") {
+		header := strings.Split(line, ": ")
+		if len(header) < 2 {
+			continue
+		}
+		headers.Add(header[0], header[1])
+	}
+	return headers
+}
+
+func mapRawData(rec analytics.AnalyticsRecord) (http.Request, http.Response, map[string]string) {
+	var req [3]string
+	var res [3]string
+	tykFields := [6]string{
+		"API-ID",
+		"API-Key",
+		"API-Name",
+		"API-Version",
+		"Oauth-ID",
+		"Org-ID",
+	}
+
+	// Decode raw HTTP transaction from base64 strings
```

Comment:
@monrax what would happens if `rec.RawRequest` or `rec.RawResponse` are empty? Usually req/resp bodies are omitted in prod and just used for debugging https://tyk.io/docs/tyk-stack/tyk-pump/useful-debug-modes/#what-is-detailed-request-logging.

Wouldn't it be better to check the length of those fields first?

Reply:
As a data lake for API calls, Resurface is able to provide context-based security analysis for folks to make data-driven decisions. In order for this to happen, we need to capture the entire payload. Because of this, we assume all the raw data is passed to `mapRawData`. So, we do check for length for these fields, just in the `WriteData` function.

Having said that, is there a way to enable detailed logging just for Resurface? (like having an option similar to `omit_detailed_recording` set to `true` for all storage destinations by default, and only enabling a backend-specific option similar to `enable_detailed_recording` only for Resurface. Sort of the opposite of how detailed logging is configured in the `tyk.conf` file at the end of that documentation page you shared)

Reply:
> Having said that, is there a way to enable detailed logging just for Resurface?
@monrax I'm afraid that's not possible. Tyk gateway is not aware of what pumps are configured here, and it just moves the analytic data to Redis to be processed by Tyk Pump. 
You can enable detailed analytics per API or even per endpoint but not for X/Y data storage since it would imply that all the API analytic data is processed and stored in Redis for pump consumption.